### PR TITLE
Added SPI_SCK_PAL_CURRENT_LEVEL configuration

### PR DESCRIPTION
**Pull request by wb-Joy**
## Description

Depends on https://github.com/ChibiOS/ChibiOS-Contrib/pull/309 and subsequent update of the submodule.

The SCK pin of WB32F371XX MCU needs to increase the driving current when using SPI, so add this code.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Retargeted to `develop` due to core changes.

**Review by fauxpark — COMMENTED** [review]
`LEVEL`, not `LEVE`?

**Comment by wb-Joy** [comment]
> 

There is detection when configuring GPIO

**Comment by wb-Joy** [comment]
> `LEVEL`, not `LEVE`?
I'll fix it right away.

File: platforms/chibios/drivers/spi_master.h [hunk 1]
```diff
@@ -63,6 +63,14 @@
 #    endif
 #endif
 
+#ifndef SPI_SCK_PAL_CURRENT_LEVEL
+#    if defined(WB32F3G71xx)
```
Comment (zvecr) [inline platforms/chibios/drivers/spi_master.h:67]: The whole point of `platforms/chibios/chibios_config.h` is so this doesnt happen

File: platforms/chibios/chibios_config.h [hunk 2]
```diff
@@ -60,6 +60,10 @@
 #        define PAL_OUTPUT_TYPE_PUSHPULL PAL_WB32_OTYPE_PUSHPULL
 #        define PAL_OUTPUT_SPEED_HIGHEST PAL_WB32_OSPEED_HIGH
 #        define PAL_PUPDR_FLOATING PAL_WB32_PUPDR_FLOATING
+#        define PAL_CURRENT_LEVEL0 PAL_WB32_CURRENT_LEVEL0
```
Comment (zvecr) [inline platforms/chibios/chibios_config.h:63]: Not really an abstraction if its only implemented for one chip type

**Review by zvecr — CHANGES_REQUESTED** [review]
Not 100% sure why this would need to be configurable

File: platforms/chibios/chibios_config.h (hunk 2 above)
Comment (wb-Joy) [reply platforms/chibios/chibios_config.h:63]: What am I supposed to do about it? Do I directly add  configuration to spi_master.c? 
```
#if defined(USE_GPIOV1)
        palSetPadMode(PAL_PORT(SPI_SCK_PIN), PAL_PAD(SPI_SCK_PIN), SPI_SCK_PAL_MODE);
        palSetPadMode(PAL_PORT(SPI_MOSI_PIN), PAL_PAD(SPI_MOSI_PIN), SPI_MOSI_PAL_MODE);
        palSetPadMode(PAL_PORT(SPI_MISO_PIN), PAL_PAD(SPI_MISO_PIN), SPI_MISO_PAL_MODE);
#else
#    if defined(WB32F3G71xx)
        palSetPadMode(PAL_PORT(SPI_SCK_PIN), PAL_PAD(SPI_SCK_PIN), PAL_MODE_ALTERNATE(SPI_SCK_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST | PAL_WB32_CURRENT_LEVEL3);
#    else
        palSetPadMode(PAL_PORT(SPI_SCK_PIN), PAL_PAD(SPI_SCK_PIN), PAL_MODE_ALTERNATE(SPI_SCK_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST);
#    endif
        palSetPadMode(PAL_PORT(SPI_MOSI_PIN), PAL_PAD(SPI_MOSI_PIN), PAL_MODE_ALTERNATE(SPI_MOSI_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST);
        palSetPadMode(PAL_PORT(SPI_MISO_PIN), PAL_PAD(SPI_MISO_PIN), PAL_MODE_ALTERNATE(SPI_MISO_PAL_MODE) | PAL_OUTPUT_TYPE_PUSHPULL | PAL_OUTPUT_SPEED_HIGHEST);
#endif
```

**Comment by wb-Joy** [comment]
ChibiOS/ChibiOS-Contrib#309 has been merged.

**Comment by wb-Joy** [comment]
Should there be some progress in this PR?

**Comment by zvecr** [comment]
> Should there be some progress in this PR?

We are not updating chibios or chibios-contrib this cycle, so no, its still on hold.

**Comment by drashna** [comment]
I believe this is mostly waiting on #16251 now.  That and for a matching update for chibios-contrib.

**Comment by wb-Joy** [comment]
Could this PR be merged as soon as possible? I have a keyboard and need it. I see that Chibios-contrib has been updated.

**Comment by drashna** [comment]
Even if it's merged today, it will be merged to develop, and won't hit master until the next breaking changes cycle (end of may, I believe). 

That said: 
> Not 100% sure why this would need to be configurable

is there a specific reason for needing this?

**Comment by wb-Joy** [comment]
> 

This is because the default IO configuration is the lowest driving capacity. If the resistance used in the circuit design of some SPI devices is too large, the driving capacity will be insufficient and normal communication will not be possible.

**Comment by wb-Joy** [comment]
With the merger now in place, customers will be able to develop keyboards based on the Develop branch and merge into the Master branch in May.

**Comment by wb-Joy** [comment]
I think this PR needs some progress, it has been 5 months.

**Comment by zvecr** [comment]
> These additions do not affect the overall architecture of the application.

I would argue they do, given its an additional abstraction that is now in place. An abstraction that personally I don't think is correct for the direction of QMK.

As for timeframes, I would say resolving comments and not actually changing anything for them is what is partially slowing this down overall.

File: platforms/chibios/drivers/spi_master.h (hunk 1 above)
Comment (zvecr) [reply platforms/chibios/drivers/spi_master.h:67]: This still needs resolving.

**Comment by tzarc** [comment]
@wb-Joy can you try the alternate solution #17650 please?

**Pull request closed without merging**